drivers/apds99xx/apds99xx.c

```diff
@@ -0,0 +1,522 @@
+/*
+ * Copyright (C) 2018 Gunar Schorcht
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_apds99xx
+ * @brief       Device driver for the Broadcom APDS99XX proximity and ambient light sensor
+ * @author      Gunar Schorcht <gunar@schorcht.net>
+ * @file
+ * @{
+ */
+
+#include <string.h>
+#include <stdlib.h>
+
+#include "apds99xx_regs.h"
+#include "apds99xx.h"
+
+#include "irq.h"
+#include "log.h"
+#include "xtimer.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+#if ENABLE_DEBUG
+
+#define ASSERT_PARAM(cond) \
```

Comment:
I already noticed that you were using all these macros in other of your driver PRs and I don't think this is following our code standards.
Even if I understand their usefulness while debugging the driver, they tend to make the code hard to read (personal opinion here) and, when used in different driver implementations, duplicate the code.

I recommend to not use them and keep the checks simple. See existing drivers.

Reply:
Thanks for your review.

Which code standards do they break? These macros are not global but only local to the according files. 

Why do think that reading `ASSERT_PARAM(cond)` is harder than reading `assert(cond)`? Also, there is no code duplication since they do not generate code if `ENABLE_DEBUG` is not enabled for the module.

The advantage of using those macros is that there is only one place where changes have to be made, for example, when the format of debug outputs has to be changed.

Reply:
@aabadie Thanks for your review. I need some more clarification. I just want to understand what the problems.

> I already noticed that you were using all these macros ...

Which of these macros are you talking about exactly? I guess `ASSERT_PARAM`, `DEBUG_DEV` and `ERROR_DEV`. As you said, they are useful for debugging and they help to avoid writing down the same source code again and again.

- `ASSERT_PARAM` simply prints the condition that is not fulfilled before just getting core panic if `ENABLE_DEBUG` is set. Finally, we could also define `DEBUG_ASSERT_VERBOSE` to get the same information. I can replace them by the `assert` function without any problem if you insist on it.
- `DEBUG_DEV` and `ERROR_DEV` have a lot of advantages:
   * They guarantee a uniform format for all debug/error messages in the module including additional information.
  * Changes that need to be made, such as the format of the output, need only to be done in one place.
  * The source code lines where they are used become significantly shorter compared to using `DEBUG` or `LOG_ERROR` directly.

> when used in different driver implementations, duplicate the code.

I don't think so. If `ENABLE_DEBUG` is 0, `DEBUG_DEV` produces no code and `ASSERT_PARAM` produces only the call of the `assert` function. In case, `NDEBUG` is defined, also `ASSERT_PARAM` produces no code. `LOG_ERROR` does not generate more code than it would generate if called directly in the source code. So I see no difference or code duplication.

> I recommend to not use them and keep the checks simple.

IMHO, checking parameters is very important during the development to realize potential problems. If `NDEBUG` is defined, the parameter checks are disabled completely and produce no code and no overhead.

I hope you are not referring to the handling of return codes from I2C-based functions. From my point of view, checking the return codes from such functions is essential for the realization of fault tolerant applications. It is not a good idea to ignore errors in error-prone functions like the buses in systems with competing hardware components. PR #10526, for example, adds a lot of such error handling.

> I don't think this is following our code standards.

Which coding standard do they break? I couldn't find anything related with the exception of _global_ multiline macros. Inline functions should be used instead of global multiline macros, but this is not the use case here.

One additional comment to the `EXEC_*` macros. They simply replace a 4 lines block by 1 line in source code. IMHO, the code becomes much more readable. Using
```
    EXEC(function1(a, b, c));
    EXEC(function2(a, b, c));
    EXEC(function3(a, b, c));
```
is much better readable than 
```
    if ((res = function1(a, b, c)) != APDS99XX_OK) {
        DEBUG("[apds99xx] %s: error code %d\n", __func__, res);
        return;
    }
    if ((res = function2(a, b, c)) != APDS99XX_OK) {
        DEBUG("[apds99xx] %s: error code %d\n", __func__, res);
        return;
    }
    if ((res = function3(a, b, c)) != APDS99XX_OK) {
        DEBUG("[apds99xx] %s: error code %d\n", __func__, res);
        return;
    }
 ```
There is also no code duplication. Even more, these macros could be extended to disable the error handling completely if `NDEBUG` is defined to reduce the code to the function call.

Reply:
Ok, good point and reasonable. Too bad, without such macros, the work becomes more difficult and the code gets worse than better readable :worried: We have used such macros over years especially in embedded systems development since they provide the capability to enable or disable parts of code for optimization without using preprocessor directives extensively everywhere.

Reply:
I have changed all macros except `DEBUG_DEV` and `ERROR_DEV` because they really help to structure the output in a consistent format.